### PR TITLE
Support keeping extra columns from shapefiles.

### LINKED ISSUE #295: Retain additional columns when pulling maps or retrieving data with with_geometry=True
**Is your feature request related to a problem? Please describe.**
For my particular application, I need some of the additional fields provided when pulling in geo data. As an example, those fields can be seen in the documentation ([example](https://tigerweb.geo.census.gov/arcgis/rest/services/TIGERweb/Tracts_Blocks/MapServer/2)).

Personally, I need the `INTPTLAT` and `INTPTLON` fields, but there are additional fields other users may benefit from. That data is also already loaded in the existing code but is discarded.

**Describe the solution you'd like**
I believe an additional, optional parameter to loading data that keeps extra columns would be an easy solution and move the responsibility of dropping extra fields to the user. One potential name for the parameter could be `with_geometry_columns` that defaults to `False` to preserve the current behavior for folks not expecting the data to be returned. 

Hypothetical example:
```python
df = ced.download(
    dataset=DATASET,
    vintage=YEAR,
    download_variables=DATA_COLUMNS,
    state=STATE_CODE,
    county=COUNTY_CODE,
    tract="*",
    block="*",
    api_key=API_KEY,
    set_to_nan=cev.ALL_SPECIAL_VALUES,
    with_geometry=True,
    with_geometry_columns=True # New parameter to keep columns
)
```

**Describe alternatives you've considered**

- Using another package to pull this data
- Parsing the cached files with custom code
- Calculating my own Lat/Lon points (would cause problems if methodology is different than used in Census-provided data)

**Additional context**

I was poking around with the debugger and identified that `__shapefile_reader` is already reading in the desired data:
```
>>> gdf.columns
Index(['STATEFP', 'COUNTYFP', 'TRACTCE', 'BLOCKCE', 'GEOID', 'GEOIDFQ', 'NAME',
       'MTFCC', 'UR', 'UACE', 'FUNCSTAT', 'ALAND', 'AWATER', 'INTPTLAT',
       'INTPTLON', 'HOUSING', 'POP', 'geometry'],
      dtype='object')
```

Therefore, I believe this could be relatively easily addressed by modifying the following to keep all the columns by not downselecting:

https://github.com/censusdis/censusdis/blob/62e8d54c4923db210fb778a7ac15eeaa89701985/censusdis/impl/us_census_shapefiles.py#L437-L439

As a crude solution, I think we could just pass along the new `with_geometry_columns` variable and change the linked code to the following. If we were worried about column conflicts, we could also specify `suffixes` for the merge operation.
```python

if with_geometry_columns:
    gdf_data = gdf_shapefile.merge(
            df_data, how="right", left_on=merge_gdf_on, right_on=df_on
        )
else:
    gdf_data = gdf_shapefile[merge_gdf_on + ["geometry"]].merge(
            df_data, how="right", left_on=merge_gdf_on, right_on=df_on
        )
```

### PR DESCRIPTION
**Pull request by vengroff**
Closes #295

### CODE REVIEW

**Comment by vengroff** [comment]
@orthorombic FYI

File: tests/test_integration.py [hunk 1]
```diff
@@ -940,27 +940,5 @@ def test_contained_within_context(self):
         )
 
 
-class UsNationTestCase(unittest.TestCase):
```
Comment (vengroff) [inline tests/test_integration.py:943]: The only change here is I moved this test to `test_geo_integration.py`, which is the proper place for it.

File: tests/test_geo_integration.py [hunk 2]
```diff
@@ -1259,3 +1303,25 @@ def test_remove_water_nj(self):
 
 if __name__ == "__main__":
     unittest.main()
+
+
+class UsNationTestCase(unittest.TestCase):
```
Comment (vengroff) [inline tests/test_geo_integration.py:1308]: This was just moved in from `test_integration.py`. It belongs here, not there.

**Comment by orthorhombic** [comment]
Everything looks good here. I'm getting some slightly unexpected results when pulling data from block groups, in that there are fewer columns than I expected (I was hoping to get "INTPTLAT" and "INTPTLON"). I admit, I don't know which api censusdis is pulling the geometry data from, so this may be my misunderstanding. Below is some analysis of which new columns are coming in, a comparison to pygris, and links to the documentation I thought applied (but might not).

## For Census block groups
New block group columns added with geometry: `{'YEAR', 'ALAND', 'GEOID', 'AWATER', 'NAMELSAD', 'LSAD', 'geometry', 'AFFGEOID'}`
Additional block group columns from pygris: `{'MTFCC', 'COUNTYFP', 'STATEFP', 'BLKGRPCE', 'INTPTLAT', 'FUNCSTAT', 'INTPTLON', 'TRACTCE'}`
Census Block Groups TIGERweb API Documentation: https://tigerweb.geo.census.gov/arcgis/rest/services/TIGERweb/Tracts_Blocks/MapServer/1

## For Census blocks
New blocks columns added with geometry: `{'MTFCC', 'POP', 'UACE', 'HOUSING', 'ALAND', 'GEOID', 'INTPTLAT', 'UR', 'FUNCSTAT', 'INTPTLON', 'AWATER', 'UATYPE', 'YEAR', 'geometry'}`
Additional blocks columns from pygris: `{'COUNTYFP', 'TRACTCE', 'STATEFP', 'BLKGRPCE', 'NAMELSAD'}`
2020 Census Blocks TIGERweb API Documentation: https://tigerweb.geo.census.gov/arcgis/rest/services/TIGERweb/Tracts_Blocks/MapServer/2

## Example code I used to output the column lists above:
```python
from censusdis import states
import censusdis.data as ced
import censusdis.values as cev
import geopandas as gpd
import pandas as pd
from pygris import blocks, block_groups
import censusdis.counties.new_jersey

state_code = states.NJ
api_key = None
county_code = censusdis.counties.new_jersey.ATLANTIC


DATA_COLUMNS = {
    "GEO_ID": "GEO_ID",
    "NAME": "NAME",
    "P3_001N": "population",  # Total population
}

df_block_groups = ced.download(
    dataset="dec/pl",
    vintage=2020,
    download_variables=DATA_COLUMNS.keys(),
    state=state_code,
    county=county_code,
    tract="*",
    block_group="*",
    api_key=api_key,
    set_to_nan=cev.ALL_SPECIAL_VALUES,
    with_geometry=False,
    with_geometry_columns=True,
)
group_no_geom_colunns = df_block_groups.columns

df_block_groups = ced.download(
    dataset="dec/pl",
    vintage=2020,
    download_variables=DATA_COLUMNS.keys(),
    state=state_code,
    county=county_code,
    tract="*",
    block_group="*",
    api_key=api_key,
    set_to_nan=cev.ALL_SPECIAL_VALUES,
    with_geometry=True,
    with_geometry_columns=True,
)
group_geom_colunns = df_block_groups.columns

block_groups_new_columns = set(group_geom_colunns).difference(set(group_no_geom_colunns))
print(f"New block group columns added with geometry: {block_groups_new_columns}")


df_tiger_block_groups = block_groups(state=state_code, county=county_code, year=2020, cache=True)
tiger_block_groups_columns = set(df_tiger_block_groups.columns)

print(f"Additional block group columns from pygris: {tiger_block_groups_columns.difference(block_groups_new_columns)}")


group_tiger_columns = df_tiger_block_groups.columns

df_blocks = ced.download(
    dataset="dec/pl",
    vintage=2020,
    download_variables=DATA_COLUMNS.keys(),
    state=state_code,
    county=county_code,
    tract="*",
    block="*",
    api_key=api_key,
    set_to_nan=cev.ALL_SPECIAL_VALUES,
    with_geometry=False,
    with_geometry_columns=True,
)
blocks_no_geom_columns = df_blocks.columns

df_blocks = ced.download(
    dataset="dec/pl",
    vintage=2020,
    download_variables=DATA_COLUMNS.keys(),
    state=state_code,
    county=county_code,
    tract="*",
    block="*",
    api_key=api_key,
    set_to_nan=cev.ALL_SPECIAL_VALUES,
    with_geometry=True,
    with_geometry_columns=True,
)
blocks_geom_colunns = df_blocks.columns

blocks_new_columns = set(blocks_geom_colunns).difference(set(blocks_no_geom_columns))
print(f"New blocks columns added with geometry: {blocks_new_columns}")


df_tiger_blocks = blocks(state=state_code, county=county_code, year=2020, cache=True)

tiger_blocks_columns = set(df_tiger_block_groups.columns)

print(f"Additional blocks columns from pygris: {tiger_blocks_columns.difference(blocks_new_columns)}")
```

**Comment by vengroff** [comment]
Thanks for the detailed example. I modified it a little to give some additional color. Here is my current version:

```python
from censusdis import states
import censusdis.data as ced
import censusdis.values as cev
import geopandas as gpd
import pandas as pd
from pygris import blocks, block_groups
import censusdis.counties.new_jersey

state_code = states.NJ
api_key = None
county_code = censusdis.counties.new_jersey.ATLANTIC


DATA_COLUMNS = {
    "GEO_ID": "GEO_ID",
    "NAME": "NAME",
    "P3_001N": "population",  # Total population
}

df_block_groups = ced.download(
    dataset="dec/pl",
    vintage=2020,
    download_variables=DATA_COLUMNS.keys(),
    state=state_code,
    county=county_code,
    tract="*",
    block_group="*",
    api_key=api_key,
    set_to_nan=cev.ALL_SPECIAL_VALUES,
    with_geometry=False,
    with_geometry_columns=True,
)
group_no_geom_colunns = df_block_groups.columns

df_block_groups = ced.download(
    dataset="dec/pl",
    vintage=2020,
    download_variables=DATA_COLUMNS.keys(),
    state=state_code,
    county=county_code,
    tract="*",
    block_group="*",
    api_key=api_key,
    set_to_nan=cev.ALL_SPECIAL_VALUES,
    with_geometry=True,
    with_geometry_columns=True,
)
group_geom_colunns = df_block_groups.columns

block_groups_new_columns = set(group_geom_colunns).difference(set(group_no_geom_colunns))
print(f"New block group columns added with geometry: {block_groups_new_columns}")
print()

df_tiger_block_groups = block_groups(state=state_code, county=county_code, year=2020, cache=True)
tiger_block_groups_columns = set(df_tiger_block_groups.columns)

print(f"Blocks all columns with geometry:: {sorted(group_geom_colunns)}")
print(f"Pygris blocks columns: {sorted(tiger_block_groups_columns)}")
print()
print(f"Additional block group columns from pygris: {sorted(tiger_block_groups_columns.difference(group_geom_colunns))}")
print(f"Block group columns not found in pygris: {sorted(group_geom_colunns.difference(tiger_block_groups_columns))}")


group_tiger_columns = df_tiger_block_groups.columns

print()
print()

df_blocks = ced.download(
    dataset="dec/pl",
    vintage=2020,
    download_variables=DATA_COLUMNS.keys(),
    state=state_code,
    county=county_code,
    tract="*",
    block="*",
    api_key=api_key,
    set_to_nan=cev.ALL_SPECIAL_VALUES,
    with_geometry=False,
    with_geometry_columns=True,
)
blocks_no_geom_columns = df_blocks.columns

df_blocks = ced.download(
    dataset="dec/pl",
    vintage=2020,
    download_variables=DATA_COLUMNS.keys(),
    state=state_code,
    county=county_code,
    tract="*",
    block="*",
    api_key=api_key,
    set_to_nan=cev.ALL_SPECIAL_VALUES,
    with_geometry=True,
    with_geometry_columns=True,
)
blocks_geom_colunns = df_blocks.columns

blocks_new_columns = set(blocks_geom_colunns).difference(set(blocks_no_geom_columns))

print(f"New blocks columns added with geometry: {blocks_new_columns}")
print()

df_tiger_blocks = blocks(state=state_code, county=county_code, year=2020, cache=True)

tiger_blocks_columns = set(df_tiger_block_groups.columns)


print(f"Blocks all columns with geometry:: {sorted(blocks_geom_colunns)}")
print(f"Pygris blocks columns: {sorted(tiger_blocks_columns)}")
print()
print(f"Additional blocks columns from pygris: {sorted(tiger_blocks_columns.difference(blocks_geom_colunns))}")
print(f"Blocks columns not found in pygris: {sorted(blocks_geom_colunns.difference(tiger_blocks_columns))}")


print()
print()
print(f"Columns in blocks but not block groups (censusdis): {sorted(blocks_geom_colunns.difference(group_geom_colunns))}")
print(f"Columns in blocks but not block groups (pygris): {sorted(tiger_blocks_columns.difference(tiger_block_groups_columns))}")
```

Looking at the output of this version, I think there are two things going on:

1. `censusdis` renames some of the columns that appear in the shapefiles. You can see this in things like it returning a column `'BLOCK'` instead of `'BLKGRPCE'` or `'STATE'` instead of `'STATEFP'`. This is due to the `censusdis` philosophy of returning columns whose names match the `kwargs` geometry specifications that were originally given, like `block=` and `state=`. In looking at it, other than this renaming, I don't think we are missing anything in these examples.
2. Some fields, including the ones you really want, like `INTPTLAT` and `INPTLON` are showing up at the block level but not the block group level. Here is what I think is going on there. `censusdis` prefers to use [U.S. Census Cartographic Boundary (CB)](https://www.census.gov/geographies/mapping-files/time-series/geo/carto-boundary-file.html) files at 500k resoulution over the full TIGER files. This is because the primary reason the `with_geometry=True` field exists is to support plotting maps, which is exactly what the CB files are for. When CB files are not available for a given geography level in a given year, `censusdis` will fall back on the full TIGER files, since that is better than nothing. That is exactly what happened here. There are no CB files at the block level, probably because the full TIGER files are fine for mapping areas that small. So we fell back on the full TIGER files. Aside from the change of resolution, it appears that these files have additional fields that the CB files do not include. And that is what is causing the differences we see.

One thing I can look into is to change the code so that if you pass in `with_geometry_columns=True` it will skip over the CB files and go straight to the full TIGER. Or, maybe the other option is to add a `tiger_geometry_only=True` option that says don't get CB files whether `with_geometry_columns` is true or not. I'll have to noodle a bit on which I like more. If you have a strong opinion, let me know.

**Comment by vengroff** [comment]
Sure enough, this little change made the extra columns from the full TIGER files show up.

```
--- a/censusdis/impl/us_census_shapefiles.py
+++ b/censusdis/impl/us_census_shapefiles.py
@@ -381,9 +381,14 @@ def add_geography(
         resolution = "5m" if shapefile_geo_level == "nation" else "500k"
 
         try:
-            gdf = __shapefile_reader(query_year).try_cb_tiger_shapefile(
-                sub_scope, shapefile_geo_level, resolution=resolution
-            )
+            if with_geometry_columns:
+                gdf = __shapefile_reader(query_year).read_shapefile(
+                    sub_scope, shapefile_geo_level
+                )
+            else:
+                gdf = __shapefile_reader(query_year).try_cb_tiger_shapefile(
+                    sub_scope, shapefile_geo_level, resolution=resolution
+                )
             gdf["YEAR"] = query_year
             return gdf
         except cmap.MapException as ex:
```

Essentially, instead of trying CB first then falling back on TIGER, if `with_geometry_columns=True` is used, we don't try CB at all and just use TIGER.

Again, I'm not totally sure if this is the best thing or if I want to make it a separate are so you can also get CB with extra fields if you really want. Since there is not much there in the CBs, I'm guessing that most people like you who want something extra really need the full TIGER.

**Comment by vengroff** [comment]
FYI, @CanyonFoot, this is related to earlier work we did on #163. So if you have comments I'd be happy to hear them.

**Comment by vengroff** [comment]
`ALAND` and `AWATER` are in the CB files, so that might be a case where someone wants CB for plotting but does want those additional columns. So that argues for the additional flag to force TIGER.

**Comment by orthorhombic** [comment]
Thanks for the analysis and quick reply. I'll admit, I didn't realize there was another source of shapefiles other than the TIGER ones, so that explains my confusion. I'll also note that I have no issues with the columns being renamed. Mostly I was confused by the different set of columns.

I tend to agree with your last comment @vengroff. It seems like separate flags would provide maximum flexibility and make it explicitly clear which data source it being used. Just brainstorming, but maybe one possibility is as follows:
- `with_geometry=True`: This keeps the existing behavior and toggles whether to return the geometry or not.
- `with_geometry_columns=True`: This keeps the existing behavior of this PR and just returns whatever extra columns are available.
- `geometry_source_data`: This could have a few different values. Maybe a default of `auto` could keep the current behavior of CB with a fallback to TIGER. `TIGER` could imply TIGER only. I guess `CB` could also be done to disable the fallback, but I'm not sure how valuable that would be.

If the feeling is this feels overly complicated for users, I also like the approach you proposed where "with_geometry_columns=True is used, we don't try CB at all and just use TIGER." I'd just suggest that we then change the flag to `return_full_tiger` because it's a little more explicit (and I think it's fun). I suspect this would probably be the most common use case after the existing behavior, though I would defer to the wisdom of others. I didn't know this package existed four days ago.

Not to add scope, but it might be worthwhile to think about how changes to pulling TIGER might impact `censusdis.maps`

**Comment by vengroff** [comment]
Thanks for the detailed follow up. I think going with the `tiger_geometry_only=` option is probably the best. I like the idea of the `geometry_source_data=` if there were or were likely to be a lot of them. But CB and TIGER are the only ones I know of for now and I don't expect the Census Bureau to add any more. I think they only made CB because the TIGER ones are much higher resolution than needed to plotting maps.

I just made another commit that adds `tiger_geometry_only=` and some additional tests that show we get different columns with CB and TIGER files at the block group level.

**Comment by vengroff** [comment]
Thanks again for taking this much time to dig into the details of a project you just discovered four days ago. How did you find us? 

And of course, please raise issues for any additional feature requests or bug reports you have, and give the project a star if you like what you see.

File: censusdis/data.py [hunk 3]
```diff
@@ -424,6 +425,17 @@ def download(
     with_geometry_columns
         If `True` keep all the additional columns that come with shapefiles
         downloaded to get geometry information.
+    tiger_shapefiles_only
+        If `True` only  look for TIGER shapefiles. If `False`, first look
```
Comment (orthorhombic) [inline censusdis/data.py:429]: Extra space between `only  look`

File: censusdis/data.py [hunk 4]
```diff
@@ -605,6 +620,17 @@ def _download_remote(
...
```
Comment (orthorhombic) [inline censusdis/data.py:624]: Extra space between `only  look`

File: censusdis/data.py [hunk 5]
```diff
@@ -1212,6 +1240,17 @@ def download(
         with_geometry_columns
             If `True` keep all the additional columns that come with shapefiles
             downloaded to get geometry information.
+        tiger_shapefiles_only
+            If `True` only  look for TIGER shapefiles. If `False`, first look
```
Comment (orthorhombic) [inline censusdis/data.py:1244]: Extra space between `only  look`

**Review by orthorhombic — COMMENTED** [review]
Going with `tiger_geometry_only` makes sense to me. Though I haven't tested it locally yet, everything looks good to me. As a super minor nit, I noticed some extra white space in the `tiger_shapefiles_only` section of docstrings which I highlighted.

I'm glad to help out, and thank you for putting together such a comprehensive interface to the census data. I found this project because I was wanting to move some code from parsing local files and move to use the API instead. I just happened across it while browsing pypi for census-related packages.

**Comment by vengroff** [comment]
Thanks. I'll fix the typo, merge and push a new release to pypi tonight.

**Comment by orthorhombic** [comment]
Thank you! I was just testing out the current branch locally and all looks good.

**Comment by vengroff** [comment]
I opened up another can of worms when I checked to make sure this worked with really wide queries. The problem is that the census API only support queries of up to 50 variables at a time. So if you ask for 51 or more, we have to do multiple queries to the underlying API and merge them together. To support this new feature, we also have to merge in the geometry and the extra shapefile columns. It works in all but a really super corner case involving data sets that return multiple rows for the same geography, like CBP and that sort of thing. That case now throws an exception explaining a better strategy for dealing with this situation.

**Pull request merged**